File: app/src/main/java/com/jy/weather/activity/ChooseCityActivity.java
package com.jy.weather.activity;

import android.Manifest;
import android.content.ComponentName;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.TextView;

import com.baidu.location.BDAbstractLocationListener;
import com.baidu.location.BDLocation;
import com.baidu.location.LocationClient;
import com.baidu.location.LocationClientOption;
import com.jy.weather.JYApplication;
import com.jy.weather.R;
import com.jy.weather.adapter.CitySearchAdapter;
import com.jy.weather.databinding.ActivityChooseCityBinding;
import com.jy.weather.util.DrawableUtil;
import com.jy.weather.util.GpsUtil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

import pub.devrel.easypermissions.EasyPermissions;

public class ChooseCityActivity extends BaseActivity implements EasyPermissions.PermissionCallbacks, View.OnClickListener {

    ActivityChooseCityBinding binding;

    private String locationCity;
    private List<String> cityList;
    private List<String> searchResult;
    private String[] PERMISSIONS = {
            Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };

    LocationClient mClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStatusBarTrans();
        setStatusBarColor();
        binding = DataBindingUtil.setContentView(this, R.layout.activity_choose_city);
        binding.setClickListener(this);

        if (!EasyPermissions.hasPermissions(this, PERMISSIONS)) {
            EasyPermissions.requestPermissions(this, "您需要同意以下权限方可使用定位功能", 1, PERMISSIONS);
        } else {
            locate();
        }
        init();
    }

    private void init() {
        String spCode = JYApplication.cityDB.getCondCode();
        if (spCode != null) {
            binding.rlChooseBackground.setBackgroundResource(DrawableUtil.INSTANCE.getBackground(spCode));
        }

        binding.ivBack.setOnClickListener(view -> {
            if (binding.etSearch.getText().length() > 0) {
                binding.etSearch.setText("");
            } else if (JYApplication.cityDB.getDefaultCity() == null) {
                addCity("北京");
                finish();
            } else {
                finish();
            }
        });

        // 添加全国城市列表
        cityList = Arrays.asList(getResources().getStringArray(R.array.national_cities_list));
        searchResult = new ArrayList<>();
        CitySearchAdapter adapter = new CitySearchAdapter(ChooseCityActivity.this, searchResult);
        binding.lvSearchResult.setAdapter(adapter);

        binding.etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.length() > 0) {// 有输入时查询匹配结果
                    searchResult.clear();
                    for (String cityInfo : cityList) {
                        if (cityInfo.contains(editable)) {
                            searchResult.add(cityInfo);
                        }
                    }
                    adapter.notifyDataSetChanged();
                    binding.llSearchResult.setVisibility(View.VISIBLE);
                    binding.llHotCity.setVisibility(View.GONE);
                } else {
                    binding.llHotCity.setVisibility(View.VISIBLE);
                    binding.llSearchResult.setVisibility(View.GONE);
                }
            }
        });
    }

    private void locate() {
        if (GpsUtil.INSTANCE.isOpen(JYApplication.INSTANCE)) {
            // 百度地图自动定位
            mClient = new LocationClient(JYApplication.INSTANCE);
            mClient.registerLocationListener(new MyLocationListener());
            LocationClientOption option = new LocationClientOption();
            option.setIsNeedAddress(true);
            option.setOpenGps(true);
            mClient.setLocOption(option);
            mClient.start();
        } else {
            onLocationFailed();
            showGpsNotOpen();
        }

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
    }

    @Override
    public void onPermissionsGranted(int requestCode, @NonNull List<String> perms) {
        if (perms.size() == PERMISSIONS.length) {
            locate();
        }
    }

    @Override
    public void onPermissionsDenied(int requestCode, @NonNull List<String> perms) {
        onLocationFailed();
        showPermissionDenied();
    }

    private void addCity(String city) {
        HashSet<String> citySet = new HashSet<>(JYApplication.cityDB.getCitySet());
        citySet.add(city);
        JYApplication.cityDB.setCitySet(citySet);
        if (citySet.size() == 1) {
            JYApplication.cityDB.setDefaultCity(city);
        }

        Intent intent = new Intent(this, WeatherActivity.class);
        intent.putExtra("city", city);
        startActivity(intent);
        finish();
    }

    @Override
    public void onClick(View v) {
        if (!isNetworkAvailable()) {
            showSnackBar(v, getString(R.string.network_unavailable));
            return;
        }

        String city = "";
        switch (v.getId()) {
            case R.id.tv_location:
                if (!GpsUtil.INSTANCE.isOpen(JYApplication.INSTANCE)) {
                    showGpsNotOpen();
                    return;
                } else {
                    if (locationCity != null) {
                        city = locationCity;
                    }
                }
                break;

            default:
                city = String.valueOf(((TextView) v).getText());
        }

        addCity(city.replace("市", ""));
    }

    private void showGpsNotOpen() {
        Snackbar snackbar = Snackbar.make(binding.tvLocation,
                getResources().getString(R.string.locate_failed),
                Snackbar.LENGTH_LONG);
        snackbar.setAction("前往打开", v -> {
//            Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
//            startActivityForResult(intent, 2);
            Intent intent = new Intent("com.jy.weather");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            ComponentName comp = new ComponentName("com.huawei.systemmanager", "com.huawei.permissionmanager.ui.MainActivity");
            intent.setComponent(comp);
            startActivity(intent);
        });
        snackbar.show();
    }

    private void showPermissionDenied() {
        Snackbar snackbar = Snackbar.make(binding.tvLocation,
                getResources().getString(R.string.locate_failed),
                Snackbar.LENGTH_LONG);
        snackbar.setAction("前往打开", v -> {
            Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
            startActivityForResult(intent, 1);
        });
        snackbar.show();
    }

    private void onLocationSuccessful(String city) {
        binding.tvLocation.setText(city);
        locationCity = city.replace("市", "")
                .replace("区", "")
                .replace("县", "");
    }

    private void onLocationFailed() {
        binding.tvLocation.setText(getResources().getString(R.string.locate_unknown));
    }

    private class MyLocationListener extends BDAbstractLocationListener {

        @Override
        public void onReceiveLocation(BDLocation location) {
            mClient.stop();

            String city = location.getCity();
            if (city == null) {
                onLocationFailed();
            } else {
                onLocationSuccessful(city);
            }
        }
    }
}

File: app/src/main/java/com/jy/weather/activity/ChooseCityActivity.kt
package com.jy.weather.activity

import android.Manifest
import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import androidx.databinding.DataBindingUtil
import androidx.databinding.Observable
import androidx.recyclerview.widget.GridLayoutManager
import com.jy.weather.JYApplication
import com.jy.weather.R
import com.jy.weather.adapter.ChooseCityAdapter
import com.jy.weather.adapter.CitySearchAdapter
import com.jy.weather.databinding.ActivityChooseCityBinding
import com.jy.weather.entity.Location
import com.jy.weather.navigator.ChooseCityNavigator
import com.jy.weather.util.AlertDialogUtil
import com.jy.weather.util.PermissionUtil
import com.jy.weather.util.SnackbarUtil
import com.jy.weather.viewmodel.ChooseCityViewModel

class ChooseCityActivity : BaseActivity(), ChooseCityNavigator {

    companion object {
        private const val PERMISSION = Manifest.permission.ACCESS_FINE_LOCATION
        private const val LOCATION_REQUEST_CODE = 0
    }

    private lateinit var binding: ActivityChooseCityBinding
    private lateinit var viewModel: ChooseCityViewModel
    private lateinit var snackbarCallback: Observable.OnPropertyChangedCallback
    private val hotCities = listOf(
        "未知" to "",
        "北京市" to "101010100",
        "天津市" to "101030100",
        "上海市" to "101020100",
        "重庆市" to "101040100",
        "沈阳市" to "101070101",
        "大连市" to "101070201",
        "长春市" to "101060101",
        "哈尔滨市" to "101050101",
        "郑州市" to "101180101",
        "武汉市" to "101200101",
        "长沙市" to "101250101",
        "广州市" to "101280101",
        "深圳市" to "101280601",
        "南京市" to "101190101",
        "杭州市" to "101210101",
        "东莞市" to "101281601",
        "成都市" to "101270101",
        "青岛市" to "101120201",
        "苏州市" to "101190401",
        "厦门市" to "101230201"
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this, R.layout.activity_choose_city)
        viewModel = ChooseCityViewModel()
        binding.viewModel = viewModel

        setupListener()

        setupSnackbarCallback()

        requestPermission()
    }

    private fun setupListener() {
        // 返回图标点击事件
        binding.ivBack.setOnClickListener {
            when {
                binding.etSearch.text.isNotEmpty() -> binding.etSearch.setText("")
                JYApplication.cityDB.defaultCity == null -> {// 默认城市北京
                    startWeatherActivity(hotCities[1].first, hotCities[1].second)
                    finish()
                }

                else -> finish()
            }
        }

        binding.rvHotCity.also {
            it.layoutManager = GridLayoutManager(this, 3)
            it.adapter = ChooseCityAdapter(hotCities) { city, cityId ->
                viewModel.onCityClick(city, cityId)
            }
        }

        binding.lvSearchResult.also {
            it.adapter = CitySearchAdapter(this)
            it.setOnItemClickListener { _, _, index, _ ->
                viewModel.onSearchResultItemClick(index)
            }
        }
    }

    private fun setupSnackbarCallback() {
        snackbarCallback = object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable?, propertyId: Int) {
                val snackbarObj = viewModel.snackbarObj.get() ?: return
                SnackbarUtil.showSnackBar(
                    window.decorView,
                    snackbarObj.text,
                    snackbarObj.action,
                    snackbarObj.listener
                )
            }
        }
        viewModel.snackbarObj.addOnPropertyChangedCallback(snackbarCallback)
    }

    private fun requestPermission() =
        PermissionUtil.checkPermissionAndRequest(
            this,
            PERMISSION,
            {
                viewModel.onPermissionGranted()
            },
            {
                showPermissionHintDialog()
            }
        )

    private fun showPermissionHintDialog() =
        AlertDialogUtil.showDialog(
            this,
            resources.getString(R.string.request_location_permission),
            {
                PermissionUtil.requestPermission(this, PERMISSION)
            },
            {
                viewModel.onPermissionDenied()
            }
        )

    override fun onResume() {
        super.onResume()

        viewModel.start(this)
    }

    override fun onDestroy() {
        viewModel.snackbarObj.removeOnPropertyChangedCallback(snackbarCallback)
        super.onDestroy()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        PermissionUtil.onPermissionResult(
            grantResults,
            {
                viewModel.onPermissionGranted()
            },
            {
                viewModel.onPermissionDenied()
            }
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == LOCATION_REQUEST_CODE) {
            viewModel.locate()
        }
    }

    override fun getActivity() = this

    override fun startOpenGpsActivity() =
        startActivityForResult(
            Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS),
            LOCATION_REQUEST_CODE
        )

    override fun startWeatherActivity(city: String?, cityId: String, location: Location?) {
        startActivity(Intent(this, WeatherActivity::class.java).apply {
            putExtra("city", city)
            putExtra("cityId", cityId)
            if (location != null) {
                putExtra("location", location)
            }
        })
        finish()
    }
}
